src/Ng/Jobs/LightningJob.cs

```diff
@@ -492,36 +516,257 @@ private static Task UpdateCursorFileAsync(string cursorFileName, int startIndex,
+                    _job.TelemetryService,
                     CancellationToken.None);
+
+                if (_job._storageFactories.SemVer2StorageFactory != null)
+                {
+                    await RegistrationMaker.ProcessAsync(
+                        new RegistrationKey(_packageId),
+                        _sortedGraphs,
+                        _job._storageFactories.SemVer2StorageFactory,
+                        new Uri(_job._contentBaseAddress),
+                        new Uri(_job._galleryBaseAddress),
+                        RegistrationCollector.PartitionSize,
+                        RegistrationCollector.PackageCountThreshold,
+                        _job._forceIconsFromFlatContainer,
+                        _job.TelemetryService,
+                        CancellationToken.None);
+                }
+            }
+        }
+
+        private class JsonStrike : IStrike
+        {
+            private readonly ICatalogClient _catalogClient;
+            private readonly IRegistrationUpdater _updater;
+            private readonly string _packageId;
+            private readonly List<string> _urls;
+
+            public JsonStrike(
+                ICatalogClient catalogClient,
+                IRegistrationUpdater updater,
+                string packageId)
+            {
+                _catalogClient = catalogClient;
+                _updater = updater;
+                _packageId = packageId;
+                _urls = new List<string>();
+            }
+
+            public async Task ProcessCatalogLeafUrlAsync(string url)
+            {
+                _urls.Add(url);
+                if (_urls.Count >= 127)
+                {
+                    await FinishAsync();
+                }
+            }
+
+            public async Task FinishAsync()
+            {
+                if (!_urls.Any())
+                {
+                    return;
+                }
+
+                // Download all of the leaves.
+                var urlsToDownload = new ConcurrentBag<string>(_urls);
+                var leaves = new ConcurrentBag<CatalogLeaf>();
+                await ParallelAsync.Repeat(
+                    async () =>
+                    {
+                        await Task.Yield();
+                        while (urlsToDownload.TryTake(out var url))
+                        {
+                            var leaf = await _catalogClient.GetLeafAsync(url);
+                            leaves.Add(leaf);
+                        }
+                    });
+
+                // Build the input to hive updaters.
+                var entries = new List<CatalogCommitItem>();
+                var entryToLeaf = new Dictionary<CatalogCommitItem, PackageDetailsCatalogLeaf>(
+                    ReferenceEqualityComparer<CatalogCommitItem>.Default);
+                foreach (var leaf in leaves)
+                {
+                    if (!(leaf.IsPackageDelete() ^ leaf.IsPackageDetails()))
```

Comment:
Is that a complicated way to write `leaf.IsPackageDelete() == leaf.IsPackageDetails()`?

Reply:
😅 yes. Your's is much better!

_"but mine uses a carrot!"_